### PR TITLE
refine isvcs healthchecks

### PR DESCRIPTION
**Pull request by tgjonestx**
Refinements to periodic health checking of iservices:
 1.  At startup, repeat attempts at healthcheck for up to 2 minutes. Fail serviced startup if the health check is still negative after 2 minutes. 
 1. After startup, if a periodic health check fails, then don't restart; just report a failed status.
 1. After startup, if an isvc exits, then restart it, UNLESS it has failed to start and run for more than 1 minute 3 times in a row (i.e. if it's starting/exiting repeatedly, then don't continue trying to restart it).

### CODE REVIEW

File: isvcs/container.go [hunk 1]
```diff
@@ -459,17 +469,61 @@ func (svc *IService) run() {
...
+				}
+				glog.Errorf("isvc %s not restarted; failed too many times", svc.Name)
+				svc.stop()
+				svc.exited = nil
 			}
 		}
 	}
 }
 
+// Return true if the service is flapping (starting/stopping repeatedly).
```
Comment (wplu) [inline isvcs/container.go]: perhaps "isFlapping returns true if the service is flapping (starting/stopping repeatedly)."

File: isvcs/container.go [hunk 2]
```diff
@@ -504,20 +558,40 @@ func (svc *IService) checkvolumes(ctr *docker.Container) bool {
 	return true
 }
 
-func (svc *IService) healthcheck(currentTime int64) <-chan error {
+// Run the default healthcheck (if any) and the return the result.
```
Comment (wplu) [inline isvcs/container.go]: perhaps "startupHealthcheck runs the default" ...

File: isvcs/container.go [hunk 3]
```diff
@@ -504,20 +558,40 @@ func (svc *IService) checkvolumes(ctr *docker.Container) bool {
...
+// If the healthcheck fails, then this method will sleep 1 second, and then
+// repeat the healthcheck, continuing that sleep/retry pattern until
+// the healthcheck succeeds or 2 minutes has elapsed.
+//
+// An error is returned if the no healtchecks succeed in the 2 minute interval,
```
Comment (wplu) [inline isvcs/container.go]: healthchecks

File: isvcs/manager.go [hunk 4]
```diff
@@ -147,20 +147,16 @@ func (m *Manager) GetHealthStatus(name string) (dao.IServiceHealthResult, error)
...
-	//        not found, or should we return an instance of IServiceHealthResult
-	//        with HealthStatuses["running"] == a-failed-instance?
-	ctr, err := docker.FindContainer(svc.name())
-	if err != nil {
-		glog.Errorf("Could not find container for isvc %s: %s", svc.Name, err)
-		return dao.IServiceHealthResult{}, err
+	if ctr, err := docker.FindContainer(svc.name()); err == nil {
+		result.ContainerID = ctr.ID
+	} else {
+		result.ContainerID = "<none>"
```
Comment (wplu) [inline isvcs/manager.go]: perhaps empty string instead of "<none>"

File: isvcs/es.go [hunk 5]
```diff
@@ -116,14 +116,20 @@ func elasticsearchHealthCheck(port int) func() error {
...
 			}
 
 			if time.Since(lastError) > minUptime {
 				break
 			}
-			time.Sleep(time.Millisecond * 1000)
+
+			select {
+			case <-halt:
+				glog.V(1).Infof("Quit healthcheck for elasticsearch at %s", baseUrl)
```
Comment (wplu) [inline isvcs/es.go]: please print the port number

Comment (tgjonestx) [inline isvcs/es.go]: The string `baseUrl` already contains the port number

Comment (wplu) [inline isvcs/es.go]: oh cool!

**Pull request merged**